clang/runtime/dpct-rt/include/dpl_extras/algorithm.h

```diff
@@ -1746,6 +1746,87 @@ segmented_reduce_argmax(Policy &&policy, Iter1 keys_in, Iter2 keys_out,
   policy.queue().wait();
 }
 
+template <typename ExecutionPolicy, typename InputIterator,
+          typename OutputIterator1, typename OutputIterator2,
+          typename OutputIterator3>
+::std::pair<OutputIterator1, OutputIterator2> nontrivial_run_length_encode(
+    ExecutionPolicy &&policy, InputIterator input_beg, InputIterator input_end,
+    OutputIterator1 offsets_out, OutputIterator2 lengths_out,
+    OutputIterator3 num_runs) {
+  using oneapi::dpl::make_transform_iterator;
+  using oneapi::dpl::make_zip_iterator;
+  using offsets_t =
+      typename ::std::iterator_traits<OutputIterator1>::value_type;
+  using lengths_t =
+      typename ::std::iterator_traits<OutputIterator2>::value_type;
+
+  auto n = ::std::distance(input_beg, input_end);
+  // First element must be nontrivial run (start of first segment)
+  auto first_adj_it = oneapi::dpl::adjacent_find(policy, input_beg, input_end);
+  auto first_adj_idx = ::std::distance(input_beg, first_adj_it);
+  if (first_adj_it == input_end) {
+    ::std::fill(policy, num_runs, num_runs + 1, 0);
+    return ::std::make_pair(offsets_out, lengths_out);
+  }
+  auto get_prev_idx_element = [first_adj_idx](const auto &idx) {
+    auto out_idx = idx + first_adj_idx;
+    return (out_idx == 0) ? 0 : out_idx - 1;
+  };
+  auto get_next_idx_element = [first_adj_idx, n](const auto &idx) {
+    auto out_idx = idx + first_adj_idx;
+    return (out_idx == n - 1) ? n - 1 : out_idx + 1;
+  };
+  // TODO: Use shifted view to pad range once oneDPL ranges is non-experimental
+  auto left_shifted_input_beg =
+      oneapi::dpl::make_permutation_iterator(input_beg, get_prev_idx_element);
+  auto right_shifted_input_beg =
+      oneapi::dpl::make_permutation_iterator(input_beg, get_next_idx_element);
+  // Segment type for ith idx consists of zip of iterators at (i-1, i, i+1)
+  // padded at the ends
+  auto zipped_keys_beg = make_zip_iterator(
+      left_shifted_input_beg, input_beg, right_shifted_input_beg,
+      oneapi::dpl::counting_iterator<offsets_t>(0));
+  // Set flag at the beginning of new nontrivial run (ex: (0, 1, 1) -> 1)
+  auto key_flags_beg =
+      make_transform_iterator(zipped_keys_beg, [n](const auto &zipped) {
+        using ::std::get;
+        bool zero_idx_mask = get<3>(zipped) != 0;
```

Comment:
I think this `zero_idx_mask` is unnecessary.  The flag predicate only cares about the `rhs`, not the `lhs`.  Also, `reduce_by_segment` always assumes that a segment starts at the zeroth element.  To pair with that we always ensure in this algo that the first element starts a segment of a non-trivial run due to the `adjacent_find` call. 

In the end, we ignore the segment flag corresponding to the zeroth index of the key, so there is no need to special case it actually.  (If we did want it, I'd think we would want `get<3>(zipped) == 0` with an `||`, not an `*` or `&&`).
 
 We **do** want the last element special case though.  If we could pad on the end with a value we know returns a false to an `==` comparison with the last element, that would be better.  However, I don't know of a general way to do that.

Reply:
Yes, you are correct. I have removed `zero_idx_mask`.